File: ts/src/classifications/indoor.test.ts
import { describe, it, expect } from 'vitest';
import {
  indoorClassificationScores,
  calculateIndoorClassification,
  coaxIndoorGroup,
} from './indoor.js';
import { Gender, Age, Bowstyle } from './data.js';
import { agbIndoorRounds, waIndoorRounds, type Round, atTarget } from '../rounds.js';
import { ScoringSystem, cm, metres } from '../targets.js';

// ---- helpers ----

function indoorRound(name: string): Round {
  return (agbIndoorRounds().get(name) ?? waIndoorRounds().get(name))!;
}

// ---- indoorClassificationScores: ages ----

describe('indoorClassificationScores ages (portsmouth recurve male)', () => {
  const cases: [Age, number[]][] = [
    [Age.Adult,   [593, 582, 566, 546, 518, 483, 437, 378]],
    [Age.Over50,  [583, 569, 549, 522, 488, 444, 387, 316]],
    [Age.Under21, [583, 569, 549, 522, 488, 444, 387, 316]],
    [Age.Under18, [571, 552, 526, 493, 450, 395, 326, 250]],
    [Age.Under16, [555, 530, 498, 457, 403, 336, 260, 187]],
    [Age.Under15, [534, 503, 463, 411, 346, 271, 196, 134]],
    [Age.Under14, [508, 469, 419, 355, 281, 206, 141,  92]],
    [Age.Under12, [475, 426, 364, 291, 215, 149,  98,  62]],
  ];
  for (const [age, want] of cases) {
    it(age, () => {
      const scores = indoorClassificationScores(
        indoorRound('portsmouth'), Bowstyle.Recurve, Gender.Male, age, true,
      );
      expect(scores).toEqual(want);
    });
  }
});

// ---- indoorClassificationScores: genders ----

describe('indoorClassificationScores genders (portsmouth recurve female)', () => {
  const cases: [Age, number[]][] = [
    [Age.Adult,   [586, 572, 553, 528, 496, 454, 399, 331]],
    [Age.Under16, [539, 510, 472, 423, 360, 286, 211, 145]],
    [Age.Under15, [534, 503, 463, 411, 346, 271, 196, 134]],
    [Age.Under12, [475, 426, 364, 291, 215, 149,  98,  62]],
  ];
  for (const [age, want] of cases) {
    it(`female ${age}`, () => {
      const scores = indoorClassificationScores(
        indoorRound('portsmouth'), Bowstyle.Recurve, Gender.Female, age, true,
      );
      expect(scores).toEqual(want);
    });
  }
});

// ---- indoorClassificationScores: bowstyles ----

describe('indoorClassificationScores bowstyles (portsmouth male adult)', () => {
  const cases: [Bowstyle, number[]][] = [
    [Bowstyle.Compound, [594, 583, 571, 560, 549, 532, 508, 472]],
    [Bowstyle.Barebow,  [565, 549, 528, 503, 472, 433, 387, 331]],
    [Bowstyle.Longbow,  [501, 466, 423, 369, 306, 240, 178, 127]],
  ];
  for (const [bowstyle, want] of cases) {
    it(bowstyle, () => {
      const scores = indoorClassificationScores(
        indoorRound('portsmouth'), bowstyle, Gender.Male, Age.Adult, true,
      );
      expect(scores).toEqual(want);
    });
  }
});

// ---- triple faces / spot variants ----

describe('indoorClassificationScores triple faces', () => {
  const cases: [string, number[]][] = [
    ['portsmouth_triple',  [594, 583, 571, 560, 549, 532, 508, 472]],
    ['worcester_5_centre', [-9999, -9999, 300, 294, 283, 267, 246, 217]],
    ['vegas_300_triple',   [300, 297, 290, 281, 269, 252, 230, 201]],
  ];
  for (const [roundName, want] of cases) {
    it(roundName, () => {
      const scores = indoorClassificationScores(
        indoorRound(roundName), Bowstyle.Compound, Gender.Male, Age.Adult, true,
      );
      expect(scores).toEqual(want);
    });
  }
});

// ---- non-strict ----

describe('indoorClassificationScores non-strict', () => {
  it('frostbite (outdoor round) with non-strict = ok', () => {
    const frostbite: Round = {
      name: 'Frostbite',
      codename: 'frostbite',
      passes: [atTarget(36, ScoringSystem.TenZone, cm(80), metres(30), false)],
      nArrows: 36,
    };
    const want = [360, 356, 349, 339, 326, 309, 286, 256];
    const scores = indoorClassificationScores(
      frostbite, Bowstyle.Compound, Gender.Male, Age.Adult, false,
    );
    expect(scores).toEqual(want);
  });
});

// ---- errors ----

describe('indoorClassificationScores errors', () => {
  it('throws for invalid bowstyle (Traditional)', () => {
    expect(() => indoorClassificationScores(
      indoorRound('portsmouth'), Bowstyle.Traditional, Gender.Male, Age.Adult, true,
    )).toThrow();
  });
  it('throws for unrecognised round', () => {
    const customRound: Round = {
      name: 'Custom', codename: 'custom_indoor_xyz',
      passes: [atTarget(36, ScoringSystem.TenZone, cm(122), metres(70), true)],
      nArrows: 36,
    };
    expect(() => indoorClassificationScores(
      customRound, Bowstyle.Recurve, Gender.Male, Age.Adult, true,
    )).toThrow();
  });
});

// ---- calculateIndoorClassification ----

describe('calculateIndoorClassification', () => {
  const cases: [number, Age, Bowstyle, string][] = [
    [594, Age.Adult,   Bowstyle.Compound, 'I-GMB'],
    [582, Age.Over50,  Bowstyle.Recurve,  'I-MB'],
    [520, Age.Under21, Bowstyle.Barebow,  'I-B1'],
    [551, Age.Under18, Bowstyle.Recurve,  'I-B1'],
    [526, Age.Under18, Bowstyle.Recurve,  'I-B1'],
    [449, Age.Under12, Bowstyle.Compound, 'I-B2'],
    [ 40, Age.Under12, Bowstyle.Longbow,  'I-A1'],
    [ 12, Age.Under12, Bowstyle.Longbow,  'UC'],
    [  1, Age.Under12, Bowstyle.Longbow,  'UC'],
  ];
  for (const [score, age, bowstyle, want] of cases) {
    it(`score=${score} ${bowstyle} ${age} → ${want}`, () => {
      const cls = calculateIndoorClassification(
        score, indoorRound('portsmouth'), bowstyle, Gender.Male, age, true,
      );
      expect(cls).toBe(want);
    });
  }
});

describe('calculateIndoorClassification invalid score', () => {
  it('throws for negative score', () => {
    expect(() => calculateIndoorClassification(
      -1, indoorRound('portsmouth'), Bowstyle.Recurve, Gender.Male, Age.Adult, true,
    )).toThrow();
  });
});

// ---- coaxIndoorGroup ----

describe('coaxIndoorGroup', () => {
  const cases: [Bowstyle, Bowstyle][] = [
    [Bowstyle.Flatbow,         Bowstyle.Barebow],
    [Bowstyle.Traditional,     Bowstyle.Barebow],
    [Bowstyle.CompoundLimited, Bowstyle.Compound],
    [Bowstyle.CompoundBarebow, Bowstyle.Compound],
    [Bowstyle.Recurve,         Bowstyle.Recurve],
  ];
  for (const [input, wantBowstyle] of cases) {
    it(`${input} → ${wantBowstyle}`, () => {
      const cat = coaxIndoorGroup(input, Gender.Male, Age.Adult);
      expect(cat.bowstyle).toBe(wantBowstyle);
    });
  }
});

File: classifications/indoor_test.go
package classifications_test

import (
	"testing"

	"github.com/retbrown/archeryutils/classifications"
	"github.com/retbrown/archeryutils/rounds"
	"github.com/retbrown/archeryutils/targets"
)

func indoorRound(t *testing.T, name string) *rounds.Round {
	t.Helper()
	allRounds := make(map[string]*rounds.Round)
	for k, v := range rounds.AGBIndoor() {
		allRounds[k] = v
	}
	for k, v := range rounds.WAIndoor() {
		allRounds[k] = v
	}
	r, ok := allRounds[name]
	if !ok {
		t.Fatalf("indoor round %q not found", name)
	}
	return r
}

func TestIndoorClassificationScoresAges(t *testing.T) {
	tests := []struct {
		age  classifications.Age
		want []int
	}{
		{classifications.Adult, []int{593, 582, 566, 546, 518, 483, 437, 378}},
		{classifications.Over50, []int{583, 569, 549, 522, 488, 444, 387, 316}},
		{classifications.Under21, []int{583, 569, 549, 522, 488, 444, 387, 316}},
		{classifications.Under18, []int{571, 552, 526, 493, 450, 395, 326, 250}},
		{classifications.Under16, []int{555, 530, 498, 457, 403, 336, 260, 187}},
		{classifications.Under15, []int{534, 503, 463, 411, 346, 271, 196, 134}},
		{classifications.Under14, []int{508, 469, 419, 355, 281, 206, 141, 92}},
		{classifications.Under12, []int{475, 426, 364, 291, 215, 149, 98, 62}},
	}
	for _, tc := range tests {
		t.Run("portsmouth_recurve_male_"+tc.age.String(), func(t *testing.T) {
			r := indoorRound(t, "portsmouth")
			got, err := classifications.IndoorClassificationScores(r, classifications.Recurve, classifications.Male, tc.age, true)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			assertIntSliceEqual(t, got, tc.want)
		})
	}
}

func TestIndoorClassificationScoresGenders(t *testing.T) {
	tests := []struct {
		age  classifications.Age
		want []int
	}{
		{classifications.Adult, []int{586, 572, 553, 528, 496, 454, 399, 331}},
		{classifications.Under16, []int{539, 510, 472, 423, 360, 286, 211, 145}},
		{classifications.Under15, []int{534, 503, 463, 411, 346, 271, 196, 134}},
		{classifications.Under12, []int{475, 426, 364, 291, 215, 149, 98, 62}},
	}
	for _, tc := range tests {
		t.Run("portsmouth_recurve_female_"+tc.age.String(), func(t *testing.T) {
			r := indoorRound(t, "portsmouth")
			got, err := classifications.IndoorClassificationScores(r, classifications.Recurve, classifications.Female, tc.age, true)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			assertIntSliceEqual(t, got, tc.want)
		})
	}
}

func TestIndoorClassificationScoresBowstyles(t *testing.T) {
	tests := []struct {
		bowstyle classifications.Bowstyle
		want     []int
	}{
		{classifications.Compound, []int{594, 583, 571, 560, 549, 532, 508, 472}},
		{classifications.Barebow, []int{565, 549, 528, 503, 472, 433, 387, 331}},
		{classifications.Longbow, []int{501, 466, 423, 369, 306, 240, 178, 127}},
	}
	for _, tc := range tests {
		t.Run("portsmouth_male_adult_"+tc.bowstyle.String(), func(t *testing.T) {
			r := indoorRound(t, "portsmouth")
			got, err := classifications.IndoorClassificationScores(r, tc.bowstyle, classifications.Male, classifications.Adult, true)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			assertIntSliceEqual(t, got, tc.want)
		})
	}
}

func TestIndoorClassificationScoresTripleFaces(t *testing.T) {
	tests := []struct {
		round string
		want  []int
	}{
		{"portsmouth_triple", []int{594, 583, 571, 560, 549, 532, 508, 472}},
		{"worcester_5_centre", []int{-9999, -9999, 300, 294, 283, 267, 246, 217}},
		{"vegas_300_triple", []int{300, 297, 290, 281, 269, 252, 230, 201}},
	}
	for _, tc := range tests {
		t.Run(tc.round, func(t *testing.T) {
			r := indoorRound(t, tc.round)
			got, err := classifications.IndoorClassificationScores(r, classifications.Compound, classifications.Male, classifications.Adult, true)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			assertIntSliceEqual(t, got, tc.want)
		})
	}
}

func TestIndoorClassificationScoresNonStrict(t *testing.T) {
	// Frostbite: 36 arrows at 30m 80cm outdoor target — not in indoor round set
	r, err := rounds.NewRound("Frostbite", []*rounds.Pass{
		mustOutdoorPass(t, 36, targets.TenZone, targets.CM(80), targets.Metres(30)),
	})
	if err != nil {
		t.Fatalf("NewRound: %v", err)
	}
	want := []int{360, 356, 349, 339, 326, 309, 286, 256}
	got, err := classifications.IndoorClassificationScores(r, classifications.Compound, classifications.Male, classifications.Adult, false)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	assertIntSliceEqual(t, got, want)
}

func TestIndoorClassificationScoresInvalidBowstyle(t *testing.T) {
	r := indoorRound(t, "portsmouth")
	_, err := classifications.IndoorClassificationScores(r, classifications.Traditional, classifications.Male, classifications.Adult, true)
	if err == nil {
		t.Fatal("expected error for invalid bowstyle")
	}
}

func TestIndoorClassificationScoresInvalidRound(t *testing.T) {
	r, _ := rounds.NewRound("Custom", []*rounds.Pass{
		mustIndoorPass(t, 36, targets.TenZone, targets.CM(122), targets.Metres(70)),
	})
	_, err := classifications.IndoorClassificationScores(r, classifications.Recurve, classifications.Male, classifications.Adult, true)
	if err == nil {
		t.Fatal("expected error for unrecognised round")
	}
}

func TestCalculateIndoorClassification(t *testing.T) {
	tests := []struct {
		score    float64
		age      classifications.Age
		bowstyle classifications.Bowstyle
		want     string
	}{
		{594, classifications.Adult, classifications.Compound, "I-GMB"},
		{582, classifications.Over50, classifications.Recurve, "I-MB"},
		{520, classifications.Under21, classifications.Barebow, "I-B1"},
		{551, classifications.Under18, classifications.Recurve, "I-B1"},
		{526, classifications.Under18, classifications.Recurve, "I-B1"},
		{449, classifications.Under12, classifications.Compound, "I-B2"},
		{40, classifications.Under12, classifications.Longbow, "I-A1"},
		{12, classifications.Under12, classifications.Longbow, "UC"},
		{1, classifications.Under12, classifications.Longbow, "UC"},
	}
	for _, tc := range tests {
		t.Run(tc.bowstyle.String()+"_"+tc.age.String()+"_"+tc.want, func(t *testing.T) {
			r := indoorRound(t, "portsmouth")
			got, err := classifications.CalculateIndoorClassification(tc.score, r, tc.bowstyle, classifications.Male, tc.age, true)
			if err != nil {
				t.Fatalf("unexpected error: %v", err)
			}
			if got != tc.want {
				t.Errorf("got %q, want %q", got, tc.want)
			}
		})
	}
}

func TestCalculateIndoorClassificationInvalidScore(t *testing.T) {
	r := indoorRound(t, "portsmouth")
	_, err := classifications.CalculateIndoorClassification(-1, r, classifications.Recurve, classifications.Male, classifications.Adult, true)
	if err == nil {
		t.Fatal("expected error for negative score")
	}
}

func TestCoaxIndoorGroup(t *testing.T) {
	tests := []struct {
		bowstyle     classifications.Bowstyle
		wantBowstyle classifications.Bowstyle
	}{
		{classifications.Flatbow, classifications.Barebow},
		{classifications.Traditional, classifications.Barebow},
		{classifications.CompoundLimited, classifications.Compound},
		{classifications.CompoundBarebow, classifications.Compound},
		{classifications.Recurve, classifications.Recurve},
	}
	for _, tc := range tests {
		t.Run(tc.bowstyle.String(), func(t *testing.T) {
			cat := classifications.CoaxIndoorGroup(tc.bowstyle, classifications.Male, classifications.Adult)
			if cat.Bowstyle != tc.wantBowstyle {
				t.Errorf("CoaxIndoorGroup(%s).Bowstyle = %s, want %s", tc.bowstyle, cat.Bowstyle, tc.wantBowstyle)
			}
		})
	}
}

func mustIndoorPass(t *testing.T, n int, sys targets.ScoringSystem, diam, dist targets.Quantity) *rounds.Pass {
	t.Helper()
	p, err := rounds.AtTarget(n, sys, diam, dist, true)
	if err != nil {
		t.Fatalf("AtTarget: %v", err)
	}
	return p
}
